lib/utils.js

```diff
@@ -1,122 +1,133 @@
   }
-};
+  return res;
+}
 
 /**
- * Default settings to use for GitHub API requests.
- *
- * @param  {Object} `config`
- * @return {Object}
+ * Create auth string - token or Basic Auth
  */
 
-utils.defaults = function (config) {
-  return function defaults(method, path, data) {
-    var opts = extend({}, config, data);
-
-    opts.apiurl = opts.apiurl || 'https://api.github.com';
-    opts.path = createURL(path, opts.apiurl);
-    opts.method = method || 'GET';
-
-    opts.headers = opts.headers || {};
-    if (!opts.headers['accept']) {
-      opts.headers['accept'] = 'application/json';
-    }
-    if (data && !opts.headers['content-type']) {
-      opts.headers['content-type'] = 'application/json';
-    }
-    if (!opts.body && opts.method.toLowerCase() === 'put') {
-      opts.headers['content-length'] = 0;
-    }
-
-    opts['user-agent'] = opts['user-agent'] || 'github-base-nodejs';
-    opts.headers['authorization'] = createAuth(opts);
-
-    delete opts.PASSWORD;
-    delete opts.USERNAME;
-    delete opts.TOKEN;
-
+function createAuth(opts) {
+  opts.headers = opts.headers || {};
+  if (opts.token) {
+    opts.headers['authorization'] = 'token ' + opts.token;
     return opts;
-  };
-};
-
-function createURL(url, base) {
-  if (url.indexOf('//') === -1) {
-    url = base + url;
   }
-  url += (/\?/.test(url) ? '&' : '?');
-  url += (new Date()).getTime();
-  return url;
-}
 
-function unpw(opts) {
-  opts = opts || {};
-  return 'Basic ' + toBase64(opts.username + ':' + opts.password);
+  var creds = opts.username + ':' + opts.password;
+  opts.headers['authorization'] = 'Basic ' + toBase64(creds);
+
+  return opts;
 }
 
+/**
+ * Cleanup request options object
+ */
+
+function cleanup(opts) {
+  var keys = [
+    'apiurl', 'token', 'username', 'password', 'placeholders'
+  ];
+  opts.placeholders = opts.placeholders.concat(keys);
+  opts = omit(opts, opts.placeholders);
 
-function createAuth(opts) {
-  opts = lowercase(opts || {});
-  if (opts.token) return 'token ' + opts.token;
-  var creds = opts.username + ':' + opts.password;
-  return 'Basic ' + toBase64(creds);
+  return opts;
 }
 
 /**
- * Lowercase the keys in the given object.
+ * Normalize request options object,
+ * the request body and few body-related headers
  */
 
-function lowercase(obj) {
-  var res = {};
-  for (var key in obj) {
-    res[key.toLowerCase()] = obj[key];
+function normalize(opts, config) {
+  var body = omit(opts, Object.keys(config).concat([
+    'headers', 'method', 'url'
+  ]));
+  var bodyKeys = Object.keys(body);
+  var isBody = bodyKeys.length !== 0;
+
+  if (isBody) {
+    opts = omit(opts, bodyKeys);
+    opts.body = JSON.stringify(body);
+    opts.headers['content-length'] = opts.body.length;
   }
-  return res;
+  if (isBody && !opts.headers['content-type']) {
+    opts.headers['content-type'] = 'application/json';
+  }
+  if (!isBody && opts.method.toLowerCase() === 'put') {
+    opts.headers['content-length'] = 0;
+  }
+
```

Comment:
mind removing the extra newline before the `return` statements? I don't know why but I don't like it lol. makes my OCD flare up